opm/autodiff/SimulatorFullyImplicitBlackoilOutput.cpp

```diff
@@ -248,5 +250,88 @@ namespace Opm
         if ( eclWriter_ ) {
             eclWriter_->writeTimeStep(timer, state, wellState);
         }
+        // write backup file
+        if( backupfile_ )
+        {
+            int reportStep      = timer.reportStepNum();
+            int currentTimeStep = timer.currentStepNum();
+            if( (reportStep == currentTimeStep || // true for SimulatorTimer
+                 currentTimeStep == 0 || // true for AdaptiveSimulatorTimer at reportStep
+                 timer.done() ) // true for AdaptiveSimulatorTimer at reportStep
+               && lastBackupReportStep_ != reportStep ) // only backup report step once
+            {
+                // store report step
+                lastBackupReportStep_ = reportStep;
+                // write resport step number
+                backupfile_.write( (const char *) &reportStep, sizeof(int) );
+
+                const BlackoilState* boState = dynamic_cast< const BlackoilState* > (&state);
+                if( boState ) {
+                    backupfile_ << (*boState);
+                }
+                else
```

Comment:
I was going to request that you add braces here, but on second thought I think it would be just as good to cast to reference instead of pointer, as that will throw on failure and the whole if block could be removed.

One small worry: I seem to remember that you need a virtual destructor for rtti and dynamic_cast to work then compiling with gcc, otherwise it will not inject the proper info. Is this just a myth or something we need to take into account?